 # 📍 **MovieLens Recommendation System**

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns

In [ ]:
ratings = pd.read_csv('./data/ratings.csv') 
ratings.head()

,userId,movieId,rating,timestamp
0,1,1,4.0,964982703
1,1,3,4.0,964981247
2,1,6,4.0,964982224
3,1,47,5.0,964983815
4,1,50,5.0,964982931


In [25]:
movies = pd.read_csv('./data/movies.csv')
movies.head()

,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy


In [26]:
tags = pd.read_csv('./data/tags.csv')
tags.head()

,userId,movieId,tag,timestamp
0,2,60756,funny,1445714994
1,2,60756,Highly quotable,1445714996
2,2,60756,will ferrell,1445714992
3,2,89774,Boxing story,1445715207
4,2,89774,MMA,1445715200


In [27]:
links = pd.read_csv('./data/links.csv')
links.head()

,movieId,imdbId,tmdbId
0,1,114709,862.0
1,2,113497,8844.0
2,3,113228,15602.0
3,4,114885,31357.0
4,5,113041,11862.0


**Problem**: Predict how a user would rate an unseen movie. (i.e.) our target variable is rating. For the data part. *ratings.csv* is our core. *movies.csv* has genre details and *tags.csv* might also be useful. I don't we will ever need *links.csv*

**Data Inspection**

In [24]:
#What are the dimensions of the dataset? (number of rows and columns)

ratings.shape

(100836, 4)

In [18]:
#What are the data types of each column and are there any missing values?

ratings.info()

<class 'pandas.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


There are *1,00,836 rows* and *4 columns*

The *timestamp* dtype is in *int64* which is weird. - (We will look into it later)

In [23]:
#What are the summary statistics of the dataset? (mean, median, standard deviation, etc.)

ratings.describe()

,userId,movieId,rating,timestamp
count,100836.000000,100836.000000,100836.000000,1.008360e+05
mean,326.127564,19435.295718,3.501557,1.205946e+09
std,182.618491,35530.987199,1.042529,2.162610e+08
min,1.000000,1.000000,0.500000,8.281246e+08
25%,177.000000,1199.000000,3.000000,1.019124e+09
50%,325.000000,2991.000000,3.500000,1.186087e+09
75%,477.000000,8122.000000,4.000000,1.435994e+09
max,610.000000,193609.000000,5.000000,1.537799e+09


Initial guess my looking here we can say that there might be *610 users* and *1,93,609 movies* - Important Trap : That would be true only if we make sure it is sequential. Due to that fact that there are only *1,00,836 rows* then having *1,93,609 movies* is not possible. But, we can't conclude anything about userId. 

In [15]:
# Checking if the userId and movieId columns are sequantial or if there are any missing values in the sequence

print(f" Unique movies: {ratings['movieId'].nunique()} and Unique users: {ratings['userId'].nunique()}")

 Unique movies: 9724 and Unique users: 610


This confirms that we have *610 Users* and *9724 Movies*

**Target Variable Stats** : The rating columns ranges from *0.5 - 5.0* and it has **mean of 3.5** (i.e.) rating above 3.5 should be modeled as good and bad if otherwise.